sys/linux/audit.go

```diff
@@ -0,0 +1,319 @@
+	}
+
+	if err := ParseNetlinkError(ack.Data); err != NLE_SUCCESS {
+		if len(ack.Data) >= 4+12 {
+			status := &AuditStatus{}
+			if err := status.fromWireFormat(ack.Data[4:]); err == nil {
+				return nil, syscall.Errno(status.Failure)
+			}
+		}
+		return nil, err
+	}
+
+	// Get the audit_status reply message. It has the same sequence number as
+	// our original request.
+	reply, err := c.getReply(seq)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get audit status reply")
+	}
+
+	if reply.Header.Type != AuditGet {
+		return nil, errors.Errorf("unexpected reply to GET, type%d", reply.Header.Type)
+	}
+
+	replyStatus := &AuditStatus{}
+	if err := replyStatus.fromWireFormat(reply.Data); err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal reply")
+	}
+
+	return replyStatus, nil
+}
+
+// SetPortID sends a netlink message to the kernel telling it the port ID of
+// netlink listener that should receive the audit messages.
+// https://github.com/linux-audit/audit-userspace/blob/990aa27ccd02f9743c4f4049887ab89678ab362a/lib/libaudit.c#L432-L464
+func (c *AuditClient) SetPortID(pid int) error {
+	netlinkPID := uint32(pid)
+	if netlinkPID == 0 {
+		netlinkPID = c.netlink.pid
+	}
+
+	status := AuditStatus{
+		Mask:    AuditStatusEnabled | AuditStatusPID,
+		Enabled: 1,
+		PID:     netlinkPID,
+	}
+	return c.set(status)
+}
+
+// RawAuditMessage is a raw audit message received from the kernel.
+type RawAuditMessage struct {
+	MessageType uint16
+	RawData     []byte // RawData is backed by the read buffer so make a copy.
+}
+
+// Receive reads an audit message from the netlink socket. If you are going to
+// use the returned message then you should make a copy of the raw data before
+// calling receive again because the raw data is backed by the read buffer.
+func (c *AuditClient) Receive(nonBlocking bool) (*RawAuditMessage, error) {
+	msgs, err := c.netlink.Receive(nonBlocking, parseNetlinkAuditMessage)
+	if err != nil {
+		return nil, err
+	}
+
+	// ParseNetlinkAuditMessage always return a slice with 1 item.
+	return &RawAuditMessage{
+		MessageType: msgs[0].Header.Type,
+		RawData:     msgs[0].Data,
+	}, nil
+}
+
+// Close closes the AuditClient and frees any associated resources.
+func (c *AuditClient) Close() error {
+	return c.netlink.Close()
+}
+
+// getReply reads from the netlink socket and find the message with the given
+// sequence number. The caller should inspect the returned message's type,
+// flags, and error code.
+func (c *AuditClient) getReply(seq uint32) (*syscall.NetlinkMessage, error) {
+	var msgs []syscall.NetlinkMessage
+	var err error
+
+	// Retry the non-blocking read multiple times until a response is received.
+	for i := 0; i < 10; i++ {
+		msgs, err = c.netlink.Receive(true, syscall.ParseNetlinkMessage)
+		if err != nil {
+			switch err {
+			case syscall.EINTR:
+				continue
+			case syscall.EAGAIN:
+				time.Sleep(50 * time.Millisecond)
```

Comment:
Yeah, the communication between user-space and the kernel is asynchronous. So it checks for a response multiple times then finally fails after 500ms.